includes/model/class-post.php

```diff
@@ -94,20 +94,22 @@ public function generate_attachments() {
 			$max_images--;
 		}
 		// then list any image attachments
-		$query = new \WP_Query(
-			array(
-				'post_parent' => $id,
-				'post_status' => 'inherit',
-				'post_type' => 'attachment',
-				'post_mime_type' => 'image',
-				'order' => 'ASC',
-				'orderby' => 'menu_order ID',
-				'posts_per_page' => $max_images,
-			)
-		);
-		foreach ( $query->get_posts() as $attachment ) {
-			if ( ! \in_array( $attachment->ID, $image_ids, true ) ) {
-				$image_ids[] = $attachment->ID;
+		if ( $max_images > 0 ) {
```

Comment:
Max images will currently always be bigger than 0, see line 84:

```PHP
		// max images can't be negative or zero
		if ( $max_images <= 0 ) {
			$max_images = 1;
		}
```

Reply:
I would also prefer to fast fail like:

```PHP
		// max images can't be negative or zero
		if ( $max_images <= 0 ) {
			return $images;
		}

Reply:
Except that `$max_images` now _can_ be zero after it's decremented on line 94 (if the user configures 1 image).  It can't return `$images` unless it gets to the metadata-building part in the loop on line 119.

However, I just realised this code is doing the wrong thing.  If the thumbnail is also one of the first attachments, which it normally would be, it will return one too few attachments.  I'll think about this some more.

Reply:
shouldn't this simply work: https://github.com/pfefferle/wordpress-activitypub/pull/248/commits/6992fbbe226eb1753d34637cb50c10b9009d778c ???

Reply:
If `$max_images` is smaller than 1, it returns an empty array, otherwise the code collects the number of images needed.

Reply:
I think this shows it a bit better: https://github.com/pfefferle/wordpress-activitypub/pull/248/files#diff-44de75651674f44bc586c0f66e8dddf70ec9c6a4a9eb72418be87aa6c12a6be7

Reply:
you were right, I think we need both... the `if` check you added and the early return... what do you think @mexon ??

Reply:
Hey, sorry I missed this conversation, but FWIW I just tested it and it behaves correctly.  The interesting case is if it delivers the right number both when the featured image is in the first N and when it is not in the first N attachments.  And that's what it does.  Honestly, I'm not sure how that works, but it does 😳

Thanks so much for carrying the patch forward!